File: android/src/org/tmurakam/cashflow/models/DescLRUManager.java
// -*-  Mode:java; c-basic-offset:4; tab-width:4; indent-tabs-mode:t -*-

package org.tmurakam.cashflow.models;

import java.util.*;

import org.tmurakam.cashflow.ormapper.*;

public class DescLRUManager {
	public static void addDescLRU(String description, int category) {
		Date now = new Date();
		addDescLRU(description, category, now);
	}

	public static void addDescLRU(String description, int category, Date date) {
		if (description == null || description.length() == 0) {
			return;
		}

		// find desc LRU from history
		String[] param = { description };
		ArrayList<Object> ary = DescLRU.instance.find_cond("WHERE description = ?", param);

		DescLRU lru;
		if (ary.size() > 0) {
			// update date
			lru = (DescLRU)ary.get(0);
		} else {
			lru = new DescLRU();
			lru.description = description;
		}
		lru.category = category;
		lru.lastUse = date;
		lru.save();
	}

	public static ArrayList<DescLRU> getDescLRUs(int category) {
		ArrayList<Object> a;

		if (category < 0) {
			// 全検索
			a = DescLRU.instance.find_cond("ORDER BY lastUse DESC LIMIT 100", null);
		} else {
			String[] param = { Integer.toString(category) };
			a = DescLRU.instance.find_cond("WHERE category = ? ORDER BY lastUse DESC LIMIT 100", param);
		}
		ArrayList<DescLRU> ary = new ArrayList<DescLRU>();
		for (Object o : a) {
			ary.add((DescLRU)o);
		}
		return ary;
	}

	/*
	  public static void gc() {
	  ArrayList<DescLRU> ary;

	  ary = DescLRU.find_cond("ORDER BY lastUse DESC LIMIT 1 OFFSET 100", null);
	  if (ary.size() > 0) {
	  DescLRU lru = ary.get(0);
	  String[] param = { date2str(lru.date) };
	  DescLRU.delete_cond("WHERE lastUse < ?", param);
	  }
	  }
	*/
}

File: Classes/models/DescLRUManager.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2011, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

#import "DescLRUManager.h"
#import "Transaction.h"

@implementation DescLRUManager

// 旧バージョンからの移行処理
// Transaction から DescLRU を生成する
+ (void)migrate
{
    NSMutableArray *ary;
    
    ary = [self getDescLRUs:-1];
    if ([ary count] > 0) {
        return;
    }
    
    // okay, we need to migrate...
    ary = [Transaction find_all:@"ORDER BY date DESC LIMIT 100"];
    for (Transaction *t in ary) {
        [self addDescLRU:t.description category:t.category date:t.date];
    }
}

+ (void)addDescLRU:(NSString *)description category:(int)category
{
    NSDate *now = [[[NSDate alloc] init] autorelease];
    [self addDescLRU:description category:category date:now];
}

+ (void)addDescLRU:(NSString *)description category:(int)category date:(NSDate*)date
{
    if ([description length] == 0) return;

    // find desc LRU from history
    DescLRU *lru = [DescLRU find_by_description:description];

    if (lru == nil) {
        // create new LRU
        lru = [[[DescLRU alloc] init] autorelease];
        lru.description = description;
    }
    lru.category = category;
    lru.lastUse = date;
    [lru save];
}

+ (NSMutableArray *)getDescLRUs:(int)category
{
    NSMutableArray *ary;

    if (category < 0) {
        // 全検索
        ary = [DescLRU find_all:@"ORDER BY lastUse DESC LIMIT 100"];
    } else {
        dbstmt *stmt = [DescLRU gen_stmt:@"WHERE category = ? ORDER BY lastUse DESC LIMIT 100"];
        [stmt bindInt:0 val:category];
        ary = [DescLRU find_all_stmt:stmt];
    }
    return ary;
}

#if 0
+ (void)gc
{
    NSMutableArray *ary = [DescLRU find:cond:@"ORDER BY lastUse DESC LIMIT 1 OFFSET 100"];
    if ([ary count] > 0) {
        DescLRU *lru = [ary objectAtIndex:0];
        dbstmt *stmt = [[Database instance] prepare:@"DELETE FROM DescLRUs WHERE lastUse < ?"];
        [stmt bindDate:0 val:lru.date];
        [stmt step];
    }
}
#endif

@end
